### DISCUSSION TITLE
Implementing 1d Klein-Gordon Equation: Evolution operator

### DISCUSSION THREAD

**Discussion by gotencoder** [zwicker-group/py-pde#376] [Q&A]
Hi there,

I am looking to implement the 1D Klein-Gordon equation $u_{tt} - u_{xx} + V = 0$  using the py-pde package but I am having some issues, specifically as the evolution operator is a second order derivative in time. I have looked at the tutorials and examples but they all seem to have an LHS as a first order derivative in time. Is there a method to do this in py-pde?
Thanks

**Comment by david-zwicker** [comment]
Yes, the package only supports first-order derivatives in time, but you can easily introduce an auxiliary field, $f=\partial_t c$, to turn your equations into a coupled system with first-order time derivatives.

**Reply by gotencoder** [reply]
Thanks for the fast reply David.
If I understand you correctly, this would then turn the problem into something of the form
$v = u_t$
$v_t = u_{xx} - V$

Is there an example you could point me to for an implementation of such a system?
Thanks

**Reply by david-zwicker** [reply]
Well, you can simply use the `PDE` class to define the equations. If you have a specific question, you can of course ask here. It might also be nice to post a full working example in the [Show and Tell category](https://github.com/zwicker-group/py-pde/discussions/categories/show-and-tell), so future visitors can profit from it.

**Reply by gotencoder** [reply]
Hi David,
I have made a toy example of a the 1d sine-gordon equation and have created a class as you advised, however I am getting unusual behaviour when integrating the kink soliton IC. The kink is supposed to move at constant speed to the right but the evolution seems to be static. Is the setup below correct?
Please let me know if you need any further information, thanks
```
def soliton(x, t ,x0=0.0, v=0.5):
    return 4 * np.arctan(np.exp((x - x0 - v*t) / np.sqrt(1 - v**2)))

class SineGordonPDE(PDEBase):
    def __init__(self, amplitude=1.0, speed=1.0):
        super().__init__()
        self.amplitude = amplitude
        self.speed = speed

    def evolution_rate(self, state, t=0):
        u, v = state  # unpack the fields

        # compute second derivatives
        u_xx = u.gradient('natural')[0].gradient('natural')[0]

        # compute the evolution rates of u and v
        u_t = v
        v_t = u_xx - self.amplitude * np.sin(u)

        return FieldCollection([u_t, v_t])
    
    def initial_conditions_soliton(self, grid):
        x = grid.axes_coords[0]
        soliton_data = soliton(x,t=0)
        u = ScalarField(grid, soliton_data)
        v = ScalarField(grid, 0.0)
        return FieldCollection([u, v])
   
```

**Reply by david-zwicker** [reply]
Instead of writing your own class, I would start with using the `PDE` class; see for instance [this example](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_brusselator_expression.html#sphx-glr-examples-gallery-pde-brusselator-expression-py). If you still run into issues, you should provide a full working example, a description of the error, and the expected result, so I can investigate.

**Reply by gotencoder** [reply]
Hi David,
Thanks - I have now put together a full working example showing the issue, please see the code below. The expected result is for the solitary wave to move to the right at a constant speed.
On a side note I get an error if I try to use a CartesianGrid instead of a UnitGrid, though I am not entirely sure why. Please let me know if you need any further information

```
from pde import PDE, FieldCollection, PlotTracker, ScalarField, UnitGrid, CartesianGrid, MemoryStorage
import numpy as np
import matplotlib.pylab as plt


def soliton(x, t ,x0=0.0, v=0.5):
    return 4 * np.arctan(np.exp((x - x0 - v*t) / np.sqrt(1 - v**2)))

# define the PDE
eq = PDE(
    {
        "u": f"v",
        "v": f"d_dx(d_dx(u)) - sin(u)",
    }
)

# initialize state
grid = UnitGrid([64, 64])
x = grid.axes_coords[0]
soliton_data = soliton(x,t=0, x0=30)
u = ScalarField(grid, soliton_data)
v = ScalarField(grid, 0.0)
state = FieldCollection([u, v])


# simulate the pde
storage = MemoryStorage()
sol = eq.solve(state, t_range=20, dt=1e-3, tracker=storage.tracker(1.0))

#Define the grid and plot the profile 
x_grid = storage.grid.cell_coords[0][:,1] 

plt.plot(x_grid,storage.data[0][0][0],linestyle='-',ms=2,c='b',marker='o',label='IC')
plt.plot(x_grid,storage.data[20][0][0],linestyle='-',ms=2,c='r',marker='o',label='sol from solver at t = 20')

plt.show() 

```

**Reply by david-zwicker** [reply]
I can reproduce the front instability. The problem lies in the definition of your PDE, which uses two nested first-order derivatives instead of a single second-order. This leads to an accumulation of numerical errors. You can fix the problem by replacing the second PDE by `"laplace(u) - sin(u)"`.

**Reply by gotencoder** [reply]
Thanks, this did indeed fix the stability issue. Is there a way to implement a CartesianGrid with a user defined spacing as the unit grid spacing is quite large for what I would like to simulate. When I replace
```
grid = UnitGrid([64, 64])
```
by 

```
grid =  CartesianGrid([[-4*np.pi, 4*np.pi]], 128, periodic=True)
```
there seems to be some issue with the periodic boundary which can be seen if we plot the IC and the solution at the first timestep

```
plt.plot(x,storage.data[0][0],linestyle='-',ms=2,c='b',marker='o',label='IC')
plt.plot(x,storage.data[1][0],linestyle='-',ms=2,c='r',marker='o',label='sol from solver')

plt.show()
```

**Reply by david-zwicker** [reply]
Are you sure you want a periodic grid? Your initial state does not seem to be compatible with a periodic grid.

**Reply by gotencoder** [reply]
You are correct, for this specific configuration periodic boundary conditions would not be a suitable choice. If I set periodic=False, instability occurs at later times in the evolution which does not seem to agree with the theory. This can be seen in the following plot

```
plt.plot(x,storage.data[0][0],linestyle='-',ms=2,c='b',marker='o',label='IC')
plt.plot(x,storage.data[10][0],linestyle='-',ms=2,c='r',marker='o',label='sol from solver')

plt.show()
```

**Reply by david-zwicker** [reply]
I cannot run the plot command, but I'm not surprised that you see instabilities, since the default method uses explicit Euler stepping, which is only first-order accurate in time. You might get better results with an implicit stepper, but I can also image that the second-order time derivative requires some more elaborate scheme for stable solutions.

**Reply by gotencoder** [reply]
Hi David,
Is there a way I can implement such a scheme using the py-pde package that produces stable solutions?

**Reply by david-zwicker** [reply]
You could sub-class [`SolverBase`](https://py-pde.readthedocs.io/en/latest/packages/pde.solvers.base.html#pde.solvers.base.SolverBase) to implement a more specific solver. I unfortunately do not have the bandwidth to work on this myself, but I'm happy to review a pull request.
